### Problem 14 - Longest Collatz sequence

The following iterative sequence is defined for the set of positive integers:

$ n → n/2$ ($n$ is even)  
$ n → 3n + 1$ ($n$ is odd)

Using the rule above and starting with 13, we generate the following sequence:

$$ 13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1 $$
    
It can be seen that this sequence (starting at $13$ and finishing at $1$) contains $10$ terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at $1$.

Which starting number, under one million, produces the longest chain?

**NOTE:** Once the chain starts the terms are allowed to go above one million.

### Solution

Create a function that follows the Collatz sequence of any initial number all the way to $1$, keeping track of every number in the sequence along the way.

Record the number with the longest sequence and its length.

In [6]:
def collatz(n):
    sequence = [n]
    while n != 1:
        # if n is even
        if n%2 == 0:
            n = n/2
            sequence.append(int(n))
        # if n is odd
        else: 
            n = n*3 + 1
            sequence.append(int(n))
    return(sequence)

In [9]:
# initiate variables
limit = 1000000
n = 2 # starting number
longest = 1 # longest sequence
largest = 1 # number with longest sequence

# loop to determine number with the longest sequence 
while n < limit:
    b = collatz(n)
    if len(b) > longest:
        longest = len(b)
        largest = n
    n += 1
print(largest, longest)

837799 525


The answer to the starting number that produces the longest Collatz sequence is **837,799** with a length of 525.